File: examples/python/mission_control/benchmark.py
#!/usr/bin/env python3
"""
Mission Control benchmark -- Python client.

Connects to a running Mission Control server (Python or TypeScript)
and measures throughput, latency percentiles, and memory usage.

Usage:
    # Start a server (either language):
    python -m mission_control.server
    #   or
    bun run examples/typescript/missionControl/server.ts

    # Run the benchmark:
    python examples/python/mission_control/benchmark.py aster1...
"""

import asyncio
import os
import statistics
import sys
import time


def mem_mb():
    """Current process RSS in MB."""
    try:
        import resource
        return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / (1024 * 1024)
    except Exception:
        return 0.0


def percentiles(latencies_ms):
    """Compute p50, p90, p99 from a list of latencies in ms."""
    if not latencies_ms:
        return 0, 0, 0
    s = sorted(latencies_ms)
    n = len(s)
    return s[n // 2], s[int(n * 0.9)], s[int(n * 0.99)]


def fmt_lat(p50, p90, p99):
    return f"p50={p50:.2f}ms  p90={p90:.2f}ms  p99={p99:.2f}ms"


async def main():
    address = sys.argv[1] if len(sys.argv) > 1 else None
    if not address:
        print("Usage: python benchmark.py <server-address>")
        sys.exit(1)

    from aster import AsterClient

    mem_start = mem_mb()
    client = AsterClient(address=address)
    await client.connect()
    mc = client.proxy("MissionControl")

    print(f"Connected to {address[:30]}...")
    print(f"Client memory at start: {mem_start:.1f} MB")
    print(f"{'─' * 72}")
    print("  ── Dynamic proxy client ────────────────────────────────────")

    # ── Warmup ────────────────────────────────────────────────────────
    for _ in range(20):
        await mc.getStatus({"agent_id": "warmup"})

    # ── Unary: getStatus ──────────────────────────────────────────────
    n_unary = 1000
    lats = []
    t0 = time.perf_counter()
    for i in range(n_unary):
        t_call = time.perf_counter()
        await mc.getStatus({"agent_id": f"bench-{i}"})
        lats.append((time.perf_counter() - t_call) * 1000)
    elapsed = time.perf_counter() - t0
    rps = n_unary / elapsed
    p50, p90, p99 = percentiles(lats)
    print(f"  Unary (getStatus)    {rps:>8,.0f} req/s   {fmt_lat(p50, p90, p99)}")

    # ── Unary: submitLog ──────────────────────────────────────────────
    n_log = 1000
    lats = []
    t0 = time.perf_counter()
    for i in range(n_log):
        t_call = time.perf_counter()
        await mc.submitLog({
            "timestamp": time.time(),
            "level": "info",
            "message": f"bench log {i}",
            "agent_id": "bench",
        })
        lats.append((time.perf_counter() - t_call) * 1000)
    elapsed = time.perf_counter() - t0
    rps = n_log / elapsed
    p50, p90, p99 = percentiles(lats)
    print(f"  Unary (submitLog)    {rps:>8,.0f} req/s   {fmt_lat(p50, p90, p99)}")

    # ── Client streaming: ingestMetrics ───────────────────────────────
    for batch_size in [100, 1_000, 10_000]:
        async def metrics(n):
            for i in range(n):
                yield {
                    "name": "cpu.usage",
                    "value": 42.0 + (i % 100) * 0.1,
                    "timestamp": time.time(),
                }

        t0 = time.perf_counter()
        result = await mc.ingestMetrics(metrics(batch_size))
        elapsed = time.perf_counter() - t0
        mps = batch_size / elapsed
        accepted = result.get("accepted", result) if isinstance(result, dict) else result
        label = f"{batch_size:>5,}"
        print(f"  Client stream ({label})  {mps:>8,.0f} msg/s   {elapsed*1000:.1f}ms total   accepted={accepted}")

    # ── Concurrent unary ──────────────────────────────────────────────
    for concurrency in [10, 50, 100]:
        t0 = time.perf_counter()
        tasks = [
            mc.getStatus({"agent_id": f"concurrent-{i}"})
            for i in range(concurrency)
        ]
        await asyncio.gather(*tasks)
        elapsed = time.perf_counter() - t0
        rps = concurrency / elapsed
        print(f"  Concurrent ({concurrency:>3})      {rps:>8,.0f} req/s   {elapsed*1000:.1f}ms total")

    # ── Typed client ──────────────────────────────────────────────────
    print("  ── Typed client (generated from contract) ──────────────────")
    from examples.python.mission_control.services import MissionControl
    from examples.python.mission_control.types import StatusRequest, LogEntry

    typed = await client.client(MissionControl)

    # Warmup
    for _ in range(20):
        await typed.getStatus(StatusRequest(agent_id="warmup"))

    # Unary getStatus
    n_unary = 1000
    lats = []
    t0 = time.perf_counter()
    for i in range(n_unary):
        t_call = time.perf_counter()
        await typed.getStatus(StatusRequest(agent_id=f"bench-{i}"))
        lats.append((time.perf_counter() - t_call) * 1000)
    elapsed = time.perf_counter() - t0
    rps = n_unary / elapsed
    p50, p90, p99 = percentiles(lats)
    print(f"  Unary (getStatus)    {rps:>8,.0f} req/s   {fmt_lat(p50, p90, p99)}")

    # Unary submitLog
    n_log = 1000
    lats = []
    t0 = time.perf_counter()
    for i in range(n_log):
        t_call = time.perf_counter()
        await typed.submitLog(LogEntry(
            timestamp=time.time(),
            level="info",
            message=f"bench log {i}",
            agent_id="bench",
        ))
        lats.append((time.perf_counter() - t_call) * 1000)
    elapsed = time.perf_counter() - t0
    rps = n_log / elapsed
    p50, p90, p99 = percentiles(lats)
    print(f"  Unary (submitLog)    {rps:>8,.0f} req/s   {fmt_lat(p50, p90, p99)}")

    # Concurrent
    for concurrency in [10, 50, 100]:
        t0 = time.perf_counter()
        tasks = [
            typed.getStatus(StatusRequest(agent_id=f"concurrent-{i}"))
            for i in range(concurrency)
        ]
        await asyncio.gather(*tasks)
        elapsed = time.perf_counter() - t0
        rps = concurrency / elapsed
        print(f"  Concurrent ({concurrency:>3})      {rps:>8,.0f} req/s   {elapsed*1000:.1f}ms total")

    # ── Memory ────────────────────────────────────────────────────────
    mem_end = mem_mb()
    print(f"{'─' * 72}")
    print(f"  Memory: start={mem_start:.1f}MB  end={mem_end:.1f}MB  delta={mem_end - mem_start:+.1f}MB")

    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/missionControl/benchmark.ts
#!/usr/bin/env bun
/**
 * Mission Control benchmark — TypeScript client.
 *
 * Connects to a running Mission Control server (Python or TypeScript)
 * and measures throughput, latency percentiles, and memory usage.
 *
 * Usage:
 *     # Start a server (either language):
 *     python -m mission_control.server
 *     #   or
 *     bun run examples/typescript/missionControl/server.ts
 *
 *     # Run the benchmark:
 *     bun run examples/typescript/missionControl/benchmark.ts aster1...
 */

import { AsterClientWrapper } from '@aster-rpc/aster';
import { MissionControl } from './services.ts';

function memMB(): number {
  try {
    return process.memoryUsage.rss() / (1024 * 1024);
  } catch {
    return 0;
  }
}

function percentiles(latencies: number[]): { p50: number; p90: number; p99: number } {
  if (latencies.length === 0) return { p50: 0, p90: 0, p99: 0 };
  const s = [...latencies].sort((a, b) => a - b);
  const n = s.length;
  return {
    p50: s[Math.floor(n * 0.5)],
    p90: s[Math.floor(n * 0.9)],
    p99: s[Math.floor(n * 0.99)],
  };
}

function fmtLat(p: { p50: number; p90: number; p99: number }): string {
  return `p50=${p.p50.toFixed(2)}ms  p90=${p.p90.toFixed(2)}ms  p99=${p.p99.toFixed(2)}ms`;
}

const address = process.argv[2];
if (!address) {
  console.error("Usage: bun run benchmark.ts <server-address>");
  process.exit(1);
}

const memStart = memMB();
const client = new AsterClientWrapper({ address });
await client.connect();
const mc = client.proxy("MissionControl");

console.log(`Connected to ${address.slice(0, 30)}...`);
console.log(`Client memory at start: ${memStart.toFixed(1)} MB`);
console.log("─".repeat(72));
console.log("  ── Dynamic proxy client ────────────────────────────────────");

// ── Warmup ────────────────────────────────────────────────────────────────
for (let i = 0; i < 20; i++) {
  await mc.getStatus({ agent_id: "warmup" });
}

// ── Unary: getStatus ──────────────────────────────────────────────────────
const N_UNARY = 1000;
let lats: number[] = [];
let t0 = performance.now();
for (let i = 0; i < N_UNARY; i++) {
  const tc = performance.now();
  await mc.getStatus({ agent_id: `bench-${i}` });
  lats.push(performance.now() - tc);
}
let elapsed = (performance.now() - t0) / 1000;
let rps = N_UNARY / elapsed;
let p = percentiles(lats);
console.log(`  Unary (getStatus)    ${rps.toFixed(0).padStart(8)} req/s   ${fmtLat(p)}`);

// ── Unary: submitLog ──────────────────────────────────────────────────────
const N_LOG = 1000;
lats = [];
t0 = performance.now();
for (let i = 0; i < N_LOG; i++) {
  const tc = performance.now();
  await mc.submitLog({
    timestamp: Date.now() / 1000,
    level: "info",
    message: `bench log ${i}`,
    agent_id: "bench",
  });
  lats.push(performance.now() - tc);
}
elapsed = (performance.now() - t0) / 1000;
rps = N_LOG / elapsed;
p = percentiles(lats);
console.log(`  Unary (submitLog)    ${rps.toFixed(0).padStart(8)} req/s   ${fmtLat(p)}`);

// ── Client streaming: ingestMetrics ───────────────────────────────────────
for (const batchSize of [100, 1_000, 10_000]) {
  async function* metrics(n: number) {
    for (let i = 0; i < n; i++) {
      yield {
        name: "cpu.usage",
        value: 42.0 + (i % 100) * 0.1,
        timestamp: Date.now() / 1000,
      };
    }
  }

  t0 = performance.now();
  const result = await mc.ingestMetrics(metrics(batchSize)) as any;
  elapsed = (performance.now() - t0) / 1000;
  const mps = batchSize / elapsed;
  const accepted = result?.accepted ?? result;
  const label = batchSize.toLocaleString().padStart(5);
  console.log(`  Client stream (${label})  ${mps.toFixed(0).padStart(8)} msg/s   ${(elapsed * 1000).toFixed(1)}ms total   accepted=${accepted}`);
}

// ── Concurrent unary ──────────────────────────────────────────────────────
for (const concurrency of [10, 50, 100]) {
  t0 = performance.now();
  const tasks = Array.from({ length: concurrency }, (_, i) =>
    mc.getStatus({ agent_id: `concurrent-${i}` }),
  );
  await Promise.all(tasks);
  elapsed = (performance.now() - t0) / 1000;
  rps = concurrency / elapsed;
  const label = concurrency.toString().padStart(3);
  console.log(`  Concurrent (${label})      ${rps.toFixed(0).padStart(8)} req/s   ${(elapsed * 1000).toFixed(1)}ms total`);
}

// ── Typed client ──────────────────────────────────────────────────────────
console.log("  ── Typed client (generated from contract) ──────────────────");
const typed: any = await client.client(MissionControl as any);

for (let i = 0; i < 20; i++) {
  await typed.getStatus({ agent_id: "warmup" });
}

lats = [];
t0 = performance.now();
for (let i = 0; i < N_UNARY; i++) {
  const tc = performance.now();
  await typed.getStatus({ agent_id: `bench-${i}` });
  lats.push(performance.now() - tc);
}
elapsed = (performance.now() - t0) / 1000;
rps = N_UNARY / elapsed;
p = percentiles(lats);
console.log(`  Unary (getStatus)    ${rps.toFixed(0).padStart(8)} req/s   ${fmtLat(p)}`);

lats = [];
t0 = performance.now();
for (let i = 0; i < N_LOG; i++) {
  const tc = performance.now();
  await typed.submitLog({
    timestamp: Date.now() / 1000,
    level: "info",
    message: `bench log ${i}`,
    agent_id: "bench",
  });
  lats.push(performance.now() - tc);
}
elapsed = (performance.now() - t0) / 1000;
rps = N_LOG / elapsed;
p = percentiles(lats);
console.log(`  Unary (submitLog)    ${rps.toFixed(0).padStart(8)} req/s   ${fmtLat(p)}`);

for (const concurrency of [10, 50, 100]) {
  t0 = performance.now();
  const tasks = Array.from({ length: concurrency }, (_, i) =>
    typed.getStatus({ agent_id: `concurrent-${i}` }),
  );
  await Promise.all(tasks);
  elapsed = (performance.now() - t0) / 1000;
  rps = concurrency / elapsed;
  const label = concurrency.toString().padStart(3);
  console.log(`  Concurrent (${label})      ${rps.toFixed(0).padStart(8)} req/s   ${(elapsed * 1000).toFixed(1)}ms total`);
}

// ── Memory ────────────────────────────────────────────────────────────────
const memEnd = memMB();
console.log("─".repeat(72));
console.log(`  Memory: start=${memStart.toFixed(1)}MB  end=${memEnd.toFixed(1)}MB  delta=${(memEnd - memStart) >= 0 ? '+' : ''}${(memEnd - memStart).toFixed(1)}MB`);

await client.close();
